### ISSUE TITLE
Add a high-performance C level implementation of `all_equal`

### ISSUE THREAD

**Issue by edward-b-1** [python/cpython#116243]
# Feature or enhancement

### Proposal:

There are several suggested ways of implementing an `all_equal(iterator)` function. The Itertools Recipe is shown in the code block below.

```python
def all_equal(iterable):
    "Returns True if all the elements are equal to each other"
    g = groupby(iterable)
    return next(g, True) and not next(g, False)
```

There are other possible Python level implementations of an `all_equal` function. As far as I have been able to tell from my background research into this, they are all slower. In other words, this itertools recipe is the fastest known implementation.

There is another possible alternative, which can be found [here](https://docs.python.org/3/library/itertools.html). I am not sure if this version is in use by either `itertools` or `more-itertools`.

```python
def all_equal(iterable, key=None):
    "Returns True if all the elements are equal to each other."
    return len(take(2, groupby(iterable, key))) <= 1
```

It looks to be similar to the above, but with a sort circuit condition.

# Comments

Some comments both for and against:

- The implementation of an `all_equal` function could perhaps be better placed elsewhere.
- In this PR, I defaulted to putting it in the built in functions module, because this is where `all` and `any` live. But it may be better placed somewhere else.
- **It is highly likely that the implementation could be improved.** Although I can program in both C and Python, combining both together requires a deep understanding of the Python implementation models, which I do not yet have. I have managed to read just about enough information from the developers guide to understand how I think this should be implemented. But I don't doubt that I have made mistakes.
- In particular I added a `PyObject_IsEqual`. Should I have really done that? I am not sure. Is the implementation of this function even correct in all cases? Again I am not sure.

- The existing itertools recipes make use of `groupby`. The operation of grouping elements of an iterable doesn't have a particularly strong connection to the operation of performing an equality test on all elements of an iterable. So, if there is demand for an `all_equal` operation, then we should probably just provide one directly.
- Doing so provides performance benefits. The implementation is in C. Part of the performance benefits come from the C implementation, and part of the performance benefits additionally come from the fact that a single function is called rather than a sequence of multiple function calls, which adds additional overhead.
- Adding an `all_equal` also provides programmer benefits. It is a convenience to have the right tool for the job rather than having to perform an internet search to find out how to do it.

# Performance

Here is some Python code which can be used to test the performance. Run it from the `cpython` directory. I do not have `more-itertools` installed globally so I chose to simply copy the implementations of any missing functions.

```python
#!./python

# test_all_equal_performance.py

from itertools import groupby, islice

def take(n, iterable):
    '''
    https://more-itertools.readthedocs.io/en/stable/_modules/more_itertools/recipes.html#take
    '''
    return list(islice(iterable, n))

def itertools_all_equal_1(iterable, key=None):
    '''
    https://docs.python.org/3/library/itertools.html
    '''
    "Returns True if all the elements are equal to each other."
    return len(take(2, groupby(iterable, key))) <= 1

def itertools_all_equal_2(iterable):
    '''
    https://more-itertools.readthedocs.io/en/stable/_modules/more_itertools/recipes.html#all_equal
    '''
    """
    Returns ``True`` if all the elements are equal to each other.

        >>> all_equal('aaaa')
        True
        >>> all_equal('aaab')
        False

    """
    g = groupby(iterable)
    return next(g, True) and not next(g, False)


def baseline(iterable):
    for _ in iterable:
        pass


from time import time
def main():

    iterable = [1] * 10**8 + [2]

    candidate_functions = [
        baseline,
        all_equal,
        itertools_all_equal_1,
        itertools_all_equal_2,
    ]

    for _ in range(5):
        for f in candidate_functions:
            t0 = time()
            print(f(iterable), f'{(time()-t0)*1e3:.1f} ms ', f.__name__)
        print(f'')


if __name__ == '__main__':
    main()
```

Here are some performance figures, obtained from both the debug and release builds.

```sh
# debug build command line
./configure --with-pydebug
make -s -j8
make test
./test_all_equal_performance.py
```

```sh
# release build command line
./configure --enable-optimizations --with-lto
make -s -j8
make test
./test_all_equal_performance.py
```

**Debug Performance:**

```
None 912.4 ms  baseline
False 833.9 ms  all_equal
False 1229.8 ms  itertools_all_equal_1
False 1264.8 ms  itertools_all_equal_2

None 1097.5 ms  baseline
False 850.6 ms  all_equal
False 1207.6 ms  itertools_all_equal_1
False 1210.2 ms  itertools_all_equal_2

None 1059.9 ms  baseline
False 838.3 ms  all_equal
False 1220.5 ms  itertools_all_equal_1
False 1238.5 ms  itertools_all_equal_2

None 1066.1 ms  baseline
False 858.0 ms  all_equal
False 1230.9 ms  itertools_all_equal_1
False 1215.4 ms  itertools_all_equal_2

None 1065.1 ms  baseline
False 840.4 ms  all_equal
False 1252.0 ms  itertools_all_equal_1
False 1240.2 ms  itertools_all_equal_2
```

**Release Performance:**

```
None 337.3 ms  baseline
False 420.6 ms  all_equal
False 695.5 ms  itertools_all_equal_1
False 698.3 ms  itertools_all_equal_2

None 406.6 ms  baseline
False 417.1 ms  all_equal
False 676.0 ms  itertools_all_equal_1
False 683.4 ms  itertools_all_equal_2

None 396.2 ms  baseline
False 417.0 ms  all_equal
False 681.2 ms  itertools_all_equal_1
False 709.5 ms  itertools_all_equal_2

None 388.0 ms  baseline
False 435.2 ms  all_equal
False 697.6 ms  itertools_all_equal_1
False 694.0 ms  itertools_all_equal_2

None 413.7 ms  baseline
False 425.9 ms  all_equal
False 698.8 ms  itertools_all_equal_1
False 682.3 ms  itertools_all_equal_2
```

Looking at the release build numbers, the C implementation is about 65 % faster.

### Has this already been discussed elsewhere?

No response given

### Linked PRs
* gh-116244

**Comment by gaogaotiantian** [comment]
Wow so there are a couple of things to mention here.

First of all, I believe you cheated with `_Py_BaseObject_RichCompare` instead of `PyObject_RichCompare`. The former basically compares the pointers which is not what you want. You probably can easily break your test with some non-interned strings, which you did not have in your test suite - another thing that is not ideal, a very limited test case.

Secondly, you are doing a much larger change than you thought - you added a **builtin** function and you add a new C API for `PyObject`! I don't think you have a very good chance to merge this as it is. To be honest I don't think `all_equal` is that much used compared to `any` and `all` - it does not make it to `itertools` and you want it to be a builtin!

So my suggestion would be - try to replace the `_Py_BaseObject_RichCompare` with `PyObject_RichCompare` and see how's the benchmark, maybe you'll find the difference is too small and the idea has no base anymore. If it's still significant, we can talk about whether this is useful enough to do a separate C implementation and where we should expose it.

**Comment by terryjreedy** [comment]
The new feature issue template begins 
> You'll need to demonstrate widespread support for your idea among the community.
Major feature proposals should generally be discussed on [Discourse](https://discuss.python.org/c/ideas/6) before opening a GitHub issue. Wait until it's clear that most people support your idea before filling in this form.

As Tian Gao said, a new builtin function *is* a major new feature.  There is an unlimited supply of possible additions, and we almost never add one that can be done in a line or two lest we bloat the module further.

I cannot immediate think of any use cases where the time would matter; but that should be discussed on a Discourse thread.

**Comment by ericvsmith** [comment]
Since (as noted) new features should be discussed on d.p.o. (discuss.python.org) first, I'm going to close this issue. It can be re-opened if the decision is to move forward with adding the feature.

And my personal opinion is that this would require a PEP, but maybe the d.p.o. discussion will say otherwise.

**Comment by edward-b-1** [comment]
> The new feature issue template begins
> 
> > You'll need to demonstrate widespread support for your idea among the community.
> > Major feature proposals should generally be discussed on [Discourse](https://discuss.python.org/c/ideas/6) before opening a GitHub issue. Wait until it's clear that most people support your idea before filling in this form.
> 
> As Tian Gao said, a new builtin function _is_ a major new feature. There is an unlimited supply of possible additions, and we almost never add one that can be done in a line or two lest we bloat the module further.
> 
> I cannot immediate think of any use cases where the time would matter; but that should be discussed on a Discourse thread.

There was a recent discussion relating to a similar, but different issue - in this case it was about adding a new form of the `map` function which would enable `operator.eq` to be used to implement `all_equal`.

I initially intended to try and implement this but realized it would be substantially more complicated, and I thought an easier target would be a good starting point, rather than trying to attempt something more complex.

I do still think having this alternative form of `map` would be useful, as it is more general than `all_equal`. Although I must admit I can't see any obvious use cases for it, except for implementing `all_equal`. That said, "more generalized things" do tend to become more useful as people realize what the tool can be used for. A specific implementation for `all_equal`, although faster, really is just for a single purpose, like `all` and `any`.

I think that probably both should be added.

Reactions to the alternative form of `map` were mixed. I suspect most people thought something along the lines of *if you want to add a high performance way of computing whether or not all elements of an iterable are equal, why not do that directly, rather than adding this somewhat obscure map thing which we don't really understand what it does*. Fair enough.

**Comment by terryjreedy** [comment]
The possibility of needing a PEP occurred to me also, but I don't remember the criterion, and discussion is needed first anyway.

**Comment by hugovk** [comment]
For reference, the discussion is at https://discuss.python.org/t/itertools-all-equal/47434.

**Comment by edward-b-1** [comment]
> First of all, I believe you cheated with `_Py_BaseObject_RichCompare` instead of `PyObject_RichCompare` (Okay you can probably use `PyObject_RichCompareBool` which gives you similar performance as `_Py_BaseObject_RichCompare`, but the benchmark should also include non-identical, but equal cases). The former basically compares the pointers which is not what you want. You probably can easily break your test with some non-interned strings, which you did not have in your test suite - another thing that is not ideal, a very limited test case.

I wouldn't say I cheated, but it is very possible that I am calling the wrong function. I am not familiar with the core of the Python implementation code (yet).

I tested on a dictionary object and you are indeed correct. It produces the wrong result. I will look to see if I can understand your suggestion and make the required change.

> Secondly, you are doing a much larger change than you thought - you added a **builtin** function and you add a new C API for `PyObject`! I don't think you have a very good chance to merge this as it is. To be honest I don't think `all_equal` is that much used compared to `any` and `all` - it does not make it to `itertools` and you want it to be a builtin!

Yes I agree, it should probably be part of itertools.

> So my suggestion would be - try to replace the `_Py_BaseObject_RichCompare` with `PyObject_RichCompareBool`, test on more than integer (an object maybe?) and see how's the benchmark. If it's still significant, we can talk about whether this is useful enough to do a separate C implementation and where we should expose it.

Thanks, I did some testing on:

```
dictionary = {
    1: 1,
}
```

and 

```
dictionary = {
    1: 'hello',
    2: 'world',
}
```

```
class TestObject():

    def __init__(self, value:int) -> None:
        self.value = value

    def __eq__(self, __value: object) -> bool:
        if isinstance(__value, TestObject):
            return self.value == __value.value
        else:
            return False
```

Some thoughts on this:

- It is probably pointless to run tests on very large objects (like very long strings or dictionaries) because then the entire runtime is biased towards spending time comparing a very long sequence of characters. In this case, because the runtime is spending the majority of its time iterating through some string, we don't see much effect due to the changes in the way `all_equal` is implemented. (Asymptotic behavior.)
- The first dictionary object is probably a sensible test case. There is about a 10 % performance gain.
- The `TestObject` has about a 3 % performance gain.

I would guess that, looking at these figures, dictionaries have their `__eq__` method implemented in C rather than Python. Is my guess correct? If not, then what could explain the difference here?

I would also caution against searching for a "worst case" comparison. It is certainly good science to have broad range of comparisons, but to search for one case where not much performance difference is realized, and then conclude because in one demonstrable case the performance gain is negligible, therefore there is no point to add an implementation is quite the opposite.

Let me provide my opinion on the results so far:

- 50 % is obviously worth having (see below)
- 10 % is probably at the point where I would begin to say it's not obvious whether it is "worth it" or not
- Anything less than 5 % I would personally conclude is a bit pointless

I did some further modification of the implementation and got it to run in less than half the time compared to the current itertools implementation.

This **may** only apply to cases where the runtime of `PyObject_RichCompareBool` is **small** compared to the overhead of the algorithm driving the calls to `PyObject_RichCompareBool`.

In other words, an `all_equal` algorithm can be split into two parts. The "driver" part which performs the iteration, and the "equality checking" part which in this case is performed by `PyObject_RichCompareBool`, whereas in the existing `itertools` module, it is done by whatever `groupby` uses.

I checked the source code for `itertoolsmodule.c` and it looks like `groupby` might be using the same function. `PyObject_RichCompareBool`, but there are many times more lines of C code, including several function calls, before that function is reached.

This makes it unsurprising that in cases where the runtime is not dominated by `PyObject_RichCompareBool`, the performance difference is significant.

**Comment by pochmann3** [comment]
You seem to have some misconceptions about itertools.

> The Itertools Recipe is shown in the code block below.

That's not the itertools recipe. It was until a few days ago. Ok, you probably just missed the update ... is what I thought until you wrote this in the same post:

> There is another possible alternative, which can be found [here](https://docs.python.org/3/library/itertools.html).

*That* is the itertools recipe. You found it in the itertools recipes section on that page.

> I am not sure if this version is in use by either `itertools`

It isn't, otherwise it wouldn't still be a recipe.

> or `more-itertools`

From its [documentation](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.all_equal) you're just one click on "source" away from seeing what that uses.

> > it does not make it to `itertools` and you want it to be a builtin!
> 
> Yes I agree, it should probably be part of itertools.

They said it does NOT make it to `itertools`. And probably the main reason isn't even its small utility but the fact that it doesn't provide an iterator.

> the current itertools implementation

There isn't one.

> in the existing `itertools` module, it is done by whatever `groupby` uses.

Not in the module but in the module's documentation, as a recipe.

**Comment by hugovk** [comment]
@edward-b-1 @pochmann3 Please continue the discussion at https://discuss.python.org/t/itertools-all-equal/47434, thanks!
